analyzer/consensus/genesis.go

```diff
@@ -35,124 +36,62 @@ func NewGenesisProcessor(logger *log.Logger) *GenesisProcessor {
 
 // Process generates SQL statements for indexing the genesis state.
 // `nodesOverride` can be nil; if non-nil, the behavior is as if the genesis document contained that set of nodes instead of whatever it contains.
-func (mg *GenesisProcessor) Process(document *genesis.Document, nodesOverride []nodeapi.Node) ([]string, error) {
-	var queries []string
+func (mg *GenesisProcessor) Process(document *genesis.Document, nodesOverride []nodeapi.Node) (*storage.QueryBatch, error) {
+	batch := &storage.QueryBatch{}
 
-	qs, err := mg.addRegistryBackendMigrations(document, nodesOverride)
-	if err != nil {
+	if err := mg.addRegistryBackendMigrations(batch, document, nodesOverride); err != nil {
 		return nil, err
 	}
-	queries = append(queries, qs...)
 
-	qs, err = mg.addStakingBackendMigrations(document)
-	if err != nil {
+	if err := mg.addStakingBackendMigrations(batch, document); err != nil {
 		return nil, err
 	}
-	queries = append(queries, qs...)
 
-	qs = mg.addGovernanceBackendMigrations(document)
-	queries = append(queries, qs...)
+	mg.addGovernanceBackendMigrations(batch, document)
 
-	mg.logger.Info("generated genesis queries", "count", len(queries))
+	mg.logger.Info("generated genesis queries", "count", batch.Len())
 
-	return queries, nil
+	return batch, nil
 }
 
-func (mg *GenesisProcessor) addRegistryBackendMigrations(document *genesis.Document, nodesOverride []nodeapi.Node) (queries []string, err error) {
+func (mg *GenesisProcessor) addRegistryBackendMigrations(batch *storage.QueryBatch, document *genesis.Document, nodesOverride []nodeapi.Node) error {
 	// Populate entities.
-	query := `INSERT INTO chain.entities (id, address)
-VALUES
-`
-	for i, signedEntity := range document.Registry.Entities {
+	for _, signedEntity := range document.Registry.Entities {
 		var entity entity.Entity
 		if err := signedEntity.Open(registry.RegisterEntitySignatureContext, &entity); err != nil {
-			return nil, err
+			return err
 		}
 
-		query += fmt.Sprintf(
-			"\t('%s', '%s')",
+		batch.Queue(queries.ConsensusEntityUpsert,
 			entity.ID.String(),
 			staking.NewAddress(entity.ID).String(),
 		)
-		if i != len(document.Registry.Entities)-1 {
-			query += ",\n"
-		}
 	}
-	query += `
-ON CONFLICT (id) DO UPDATE SET address = EXCLUDED.address;`
```

Comment:
You mention in the description that not all `ON CONFLICT` are exactly preserved. That does worry me a little, because the behavior can be nuanced. A good test would be to run e2e with a tweaked `e2e_config_1.yml` so that it uses fast sync for half (?) of the normal block range. After fast sync, analyzers would run StateToGenesis and test our genesis processing against a prepopulated db. But that's probably out of the question because StateToGenesis is unavailable. Kinda ironic because StateToGenesis being unavailable is what made you pause you previous work and go do this PR instead :). cc @gw0 as a data point towards https://app.clickup.com/t/8693mg8z2

Reply:
added a test in #667, no regressions